scripts/release/build.sh

```diff
@@ -0,0 +1,108 @@
+#
+# This script needs to be run in a terminal with a human watching to
+# be prompted for GPG key password at a couple points.
+#
+# Externally settable env vars:
+# S3_PREFIX= where to upload build artifacts (no trailing /)
+# RSTAMP= `scripts/reverse_hex_timestamp`
+# AWS_ACCESS_KEY_ID=
+# AWS_SECRET_ACCESS_KEY=
+
+date "+build_release start %Y%m%d_%H%M%S"
+
+set -ex
+
+# Anchor our repo root reference location
+# /home/ubuntu/..
+# /home/ubuntu/go/src/github.com/algorand/go-algorand/
+#REPO_ROOT="$( cd "$(dirname "$0")" ; pwd -P )"/..
+REPO_ROOT=/home/ubuntu/go/src/github.com/algorand/go-algorand/
+#if [ -f ${REPO_ROOT}/crypto/libsodium-fork/Makefile ]; then
+#    (cd ${REPO_ROOT}/crypto/libsodium-fork && make distclean)
+#fi
+#rm -rf ${REPO_ROOT}/crypto/lib
+
+cd ${REPO_ROOT}
+export RELEASE_GENESIS_PROCESS=true
+PLATFORM=$(./scripts/osarchtype.sh)
+PLATFORM_SPLIT=(${PLATFORM//\// })
+OS=${PLATFORM_SPLIT[0]}
+ARCH=${PLATFORM_SPLIT[1]}
+BRANCH=$(./scripts/compute_branch.sh)
+export BRANCH
+CHANNEL=$(./scripts/compute_branch_channel.sh "${BRANCH}")
+export CHANNEL
+DEFAULTNETWORK=$(./scripts/compute_branch_network.sh)
+export DEFAULTNETWORK
+export PKG_ROOT=${HOME}/node_pkg
+export VARIATIONS="base"
+# tell underlying 'build' scripts we already built
+export NO_BUILD=true
+
+RSTAMP=$(scripts/reverse_hex_timestamp)
+echo RSTAMP="${RSTAMP}" > "${HOME}/rstamp"
+
+# What's my default IP address?
+# get the datacenter IP address for this EC2 host.
+# this might equivalently be gotten from `netstat -rn` and `ifconfig -a`
+DC_IP=$(curl --silent http://169.254.169.254/latest/meta-data/local-ipv4)
+if [ -z "${DC_IP}" ]; then
+    echo "ERROR: need DC_IP to be set to your local (but not localhost) IP"
+    exit 1
+fi
+
+# Update version file for this build
+if [ ! -z "${BUILD_NUMBER}" ]; then
+    echo "using externally set BUILD_NUMBER=${BUILD_NUMBER} without incrementing"
+else
+    if [ -e buildnumber.dat ]; then
+	BUILD_NUMBER=$(cat ./buildnumber.dat)
+	BUILD_NUMBER=$(( BUILD_NUMBER + 1 ))
+    else
+	BUILD_NUMBER=0
+    fi
+    echo ${BUILD_NUMBER} > ./buildnumber.dat
+    git add -A
+    git commit -m "Build ${BUILD_NUMBER}"
+fi
+FULLVERSION=$(./scripts/compute_build_number.sh -f)
+export FULLVERSION
+
+# a bash user might `source build_env` to manually continue a broken build
+cat <<EOF>"${HOME}"/build_env
+export RELEASE_GENESIS_PROCESS=${RELEASE_GENESIS_PROCESS}
+PLATFORM=${PLATFORM}
+OS=${OS}
+ARCH=${ARCH}
+export BRANCH=${BRANCH}
+export CHANNEL=${CHANNEL}
+export DEFAULTNETWORK=${DEFAULTNETWORK}
+export PKG_ROOT=${PKG_ROOT}
+export VARIATIONS=${VARIATIONS}
+RSTAMP=${RSTAMP}
+BUILD_NUMBER=${BUILD_NUMBER}
+export FULLVERSION=${FULLVERSION}
+DC_IP=${DC_IP}
+EOF
+# strip leading 'export ' for docker --env-file
+sed 's/^export //g' < "${HOME}"/build_env > "${HOME}"/build_env_docker
+
+export GOPATH=${HOME}/go
+export PATH=${HOME}/gpgbin:${GOPATH}/bin:/usr/local/go/bin:${PATH}
+
+# Build!
+scripts/configure_dev.sh
+make crypto/lib/libsodium.a
```

Comment:
why ? doesn't make build creates that automatically ?

Reply:
I'll have to check with Brian to see if there's a reason to build this.

Reply:
this _shouldn't_ be necessary, here and a couple other places, `make build` _should_ take care of building libsodium, but I guess at one point in the past this worked around a finicky build. Probably we should try without it and see if we don't need it and maybe this time we can take the time to fix it right, or comment here why it's a good workaround

Reply:
Brian is correct. At some point in the past it *was* needed.
Today, assuming that you're building on a clean environment every time, there *shouldn't* be a need to do that. ( but please do correct me if I'm wrong )

Reply:
I've found that if I don't include this right before we make the rpm package that it won't build.

For example, if I comment out the code here:
```
# definitely rebuild libsodium which could link to external C libraries
#if [ -f ${REPO_DIR}/crypto/libsodium-fork/Makefile ]; then
#    (cd ${REPO_DIR}/crypto/libsodium-fork && make distclean)
#fi
#rm -rf ${REPO_DIR}/crypto/lib
#make crypto/lib/libsodium.a

make build

RPMTMP=$(mktemp -d 2>/dev/null || mktemp -d -t "rpmtmp")
trap 'rm -rf ${RPMTMP}' 0
"${REPO_DIR}/scripts/release/helper/build_rpm.sh" "${RPMTMP}"
```
I get the following build errors:
```
crypto/lib/libsodium.a(libsodium_la-utils.o): In function `explicit_bzero':
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:83: undefined reference to `__explicit_bzero_chk'
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:83: undefined reference to `__explicit_bzero_chk'
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:83: undefined reference to `__explicit_bzero_chk'
collect2: error: ld returned 1 exit status
make: *** [buildsrc] Error 2
build_release done packaging centos 20200108_033715
```

Reply:
ok, inside the rpm build inside the centos docker container I know we need the cleanup lines. The outer ubuntu build and the inner centos build are using the same checkout to ensure they have exactly the same source version. (We might be able to get rid of the entire 'centos build' now, we should do testing that current builds built on ubuntu18 run on ubuntu16 and centos7; this should work because of Tsachi's changes to make the build fully static)

The question I had about the build scripts was can we just skip the line `make crypto/lib/libsodium.a` that shows up in a few places.